File: spikes/visionos_host/apple/Sources/DisplayList.swift
// SPDX-License-Identifier: MPL-2.0
import Foundation
import SwiftUI

enum DisplayOpcode: UInt8 {
    case fillRect = 1
    case fillRoundedRect = 2
    case strokeRoundedRect = 3
}

struct DisplayCommand: Equatable {
    let opcode: DisplayOpcode
    let rect: CGRect
    let radius: CGFloat
    let red: Double
    let green: Double
    let blue: Double
    let alpha: Double

    var color: Color {
        Color(red: red, green: green, blue: blue, opacity: alpha)
    }
}

enum DisplayListError: Error, Equatable {
    case shortHeader
    case invalidMagic
    case unsupportedHeader(version: UInt16, size: UInt16)
    case invalidLength
    case invalidOpcode(UInt8)
    case invalidValue
}

enum DisplayListDecoder {
    static let headerSize = 12
    static let commandSize = 40
    static let maximumCommandCount = 100_000

    static func decode(_ data: Data) throws -> [DisplayCommand] {
        guard data.count >= headerSize else { throw DisplayListError.shortHeader }
        guard Array(data[0..<4]) == [0x46, 0x57, 0x44, 0x4c] else {
            throw DisplayListError.invalidMagic
        }
        let version = readUInt16(data, 4)
        let encodedHeaderSize = readUInt16(data, 6)
        guard version == 1, encodedHeaderSize == headerSize else {
            throw DisplayListError.unsupportedHeader(
                version: version,
                size: encodedHeaderSize
            )
        }
        let commandCount = Int(readUInt32(data, 8))
        guard commandCount <= maximumCommandCount,
              data.count == headerSize + commandCount * commandSize else {
            throw DisplayListError.invalidLength
        }

        return try (0..<commandCount).map { index in
            let offset = headerSize + index * commandSize
            guard data[offset + 1] == 0, data[offset + 2] == 0,
                  data[offset + 3] == 0 else {
                throw DisplayListError.invalidValue
            }
            guard let opcode = DisplayOpcode(rawValue: data[offset]) else {
                throw DisplayListError.invalidOpcode(data[offset])
            }
            let values = (0..<9).map { valueIndex in
                Double(readFloat(data, offset + 4 + valueIndex * 4))
            }
            guard values.allSatisfy(\.isFinite),
                  values[2] >= 0, values[3] >= 0, values[4] >= 0,
                  values[5...8].allSatisfy({ 0...1 ~= $0 }) else {
                throw DisplayListError.invalidValue
            }
            return DisplayCommand(
                opcode: opcode,
                rect: CGRect(x: values[0], y: values[1],
                             width: values[2], height: values[3]),
                radius: values[4],
                red: values[5], green: values[6], blue: values[7],
                alpha: values[8]
            )
        }
    }

    private static func readUInt16(_ data: Data, _ offset: Int) -> UInt16 {
        UInt16(data[offset]) | (UInt16(data[offset + 1]) << 8)
    }

    private static func readUInt32(_ data: Data, _ offset: Int) -> UInt32 {
        UInt32(data[offset]) |
            (UInt32(data[offset + 1]) << 8) |
            (UInt32(data[offset + 2]) << 16) |
            (UInt32(data[offset + 3]) << 24)
    }

    private static func readFloat(_ data: Data, _ offset: Int) -> Float {
        Float(bitPattern: readUInt32(data, offset))
    }
}

File: spikes/playground_ui/lib/src/display_list.dart
// SPDX-License-Identifier: MPL-2.0
import 'dart:typed_data';

import 'models.dart';

const int _headerSize = 12;
const int _commandSize = 40;
const int _maxCommandCount = 100000;

List<DisplayCommand> decodeDisplayList(Uint8List bytes) {
  if (bytes.length < _headerSize) {
    throw const FormatException('Display list is shorter than its header.');
  }
  if (bytes[0] != 0x46 ||
      bytes[1] != 0x57 ||
      bytes[2] != 0x44 ||
      bytes[3] != 0x4c) {
    throw const FormatException('Invalid display-list magic.');
  }

  final data = ByteData.sublistView(bytes);
  final version = data.getUint16(4, Endian.little);
  final headerSize = data.getUint16(6, Endian.little);
  final commandCount = data.getUint32(8, Endian.little);
  if (version != 1 || headerSize != _headerSize) {
    throw FormatException(
      'Unsupported display-list header: version=$version size=$headerSize.',
    );
  }
  if (commandCount > _maxCommandCount) {
    throw const FormatException('Display-list command limit exceeded.');
  }
  final expectedLength = headerSize + commandCount * _commandSize;
  if (bytes.length != expectedLength) {
    throw FormatException(
      'Display-list length is ${bytes.length}; expected $expectedLength.',
    );
  }

  return List<DisplayCommand>.generate(commandCount, (index) {
    final offset = headerSize + index * _commandSize;
    if (bytes[offset + 1] != 0 ||
        bytes[offset + 2] != 0 ||
        bytes[offset + 3] != 0) {
      throw const FormatException('Reserved command bytes must be zero.');
    }
    final values = List<double>.generate(
      9,
      (valueIndex) =>
          data.getFloat32(offset + 4 + valueIndex * 4, Endian.little),
      growable: false,
    );
    if (values.any((value) => !value.isFinite)) {
      throw const FormatException('Display command contains non-finite data.');
    }
    if (values[2] < 0 ||
        values[3] < 0 ||
        values[4] < 0 ||
        values.skip(5).any((value) => value < 0 || value > 1)) {
      throw const FormatException('Display command is outside valid ranges.');
    }
    return DisplayCommand(
      opcode: DisplayOpcode.fromWire(bytes[offset]),
      x: values[0],
      y: values[1],
      width: values[2],
      height: values[3],
      radius: values[4],
      red: values[5],
      green: values[6],
      blue: values[7],
      alpha: values[8],
    );
  }, growable: false);
}
